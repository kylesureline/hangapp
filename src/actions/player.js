import database from '../firebase/firebase';

export const initializeGame = (word, guessesRemaining) => ({
  type: 'INIT_GAME',
  answer: word,
  guessesRemaining
});

export const guessLetter = (letter) => ({
  type: 'GUESS_LETTER',
  letter
});

export const updateGuessedWord = (wordArray) => ({
  type: 'UPDATE_GUESSED_WORD',
  wordArray
});

export const wrongGuess = () => ({
  type: 'WRONG_GUESS'
});

export const setPlayer = (player) => ({
  type: 'SET_PLAYER',
  player
});

export const setPlayerStats = (stats) => ({
  type: 'SET_PLAYER_STATS',
  stats
});

export const setPastGames = (pastGames) => ({
  type: 'SET_PAST_GAMES',
  pastGames
});

export const startSetPlayer = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;

    return database.ref(`players/${uid}`)
      .once('value')
      .then((snapshot) => {
        const pastGames = [];

        for(let game in snapshot.val().pastGames) {
          pastGames.push(snapshot.val().pastGames[game]);
        }

        const stats = snapshot.val().stats;

        dispatch(setPlayerStats(stats));
        dispatch(setPastGames(pastGames));
      });

  };
};
