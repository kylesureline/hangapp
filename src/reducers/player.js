const defaultGameState = {
  answer: {
    word: '',
    type: '',
    def: ''
  },
  guessedLetters: [],
  guessedWord: [],
  guessesRemaining: 0,
  stats: {
    wins: 0,
    losses: 0
  }
};

export default (state = defaultGameState, action) => {
  switch(action.type) {
    case 'GUESS_LETTER':
      return {
        ...state,
        guessedLetters: [...state.guessedLetters, action.letter]
      };
    case 'UPDATE_GUESSED_WORD':
      return {
        ...state
      };
    case 'WRONG_GUESS':
      return {
        ...state,
        guessesRemaining: state.guessesRemaining -= 1
      };
    case 'SET_PLAYER_STATS':
      return {
        ...state,
        stats: action.stats
      };
    case 'SET_PAST_GAMES':
      return {
        ...state,
        pastGames: action.pastGames
      };
    case 'SET_GUESSES_REMAINING':
      return {
        ...state,
        guessesRemaining: action.guessesRemaining
      }
    case 'CHOOSE_DEFAULT_WORD':
      return {
        ...state,
        answer: {
          word: 'hangapp',
          type: 'noun',
          def: 'A hangman game built as a React PWA by Kyle Scheuerlein'
        }
      };
    case 'CHOOSE_RANDOM_WORD':
      return {
        ...state,
        answer: action.answer
      }
    case 'ADD_WORD':
      return {
        ...state,
        pastGames: [
          ...state.pastGames,
          action.answer
        ]
      }
    case 'ADD_WIN':
      return {
        ...state,
        stats: {
          ...state.stats,
          wins: state.stats.wins + 1
        }
      }
    case 'ADD_LOSS':
      return {
        ...state,
        stats: {
          ...state.stats,
          losses: state.stats.losses + 1
        }
      }
    default:
      return state;
  }
};
