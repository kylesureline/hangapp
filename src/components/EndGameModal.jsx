import React from 'react';
import Modal from 'react-modal';
import { connect } from 'react-redux';
import { startSetPlayer } from '../actions/player';

const EndGameModal = (props) => {
  const guessesRemaining = props.player.guessesRemaining;
  const answer = props.player.answer;
  const guessedWord = props.player.guessedWord.join('');
  return (
    <Modal
      isOpen={guessesRemaining <= 0 || answer.word === guessedWord}
      onRequestClose={props.startSetPlayer}
      contentLabel="Game over!"
      className="modal"
    >
      <h3>Game over!</h3>
      <p>You {
        answer.word === guessedWord ? 'won' : 'lost'
      }!</p>
      <p>{answer.word}, {answer.type}, {answer.def}</p>
      <button onClick={props.startSetPlayer}>Play again?</button>
    </Modal>
  );
};

const mapStateToProps = (state) => ({
  player: state.player
});

const mapDispatchToProps = (dispatch) => ({
  startSetPlayer: () => dispatch(startSetPlayer())
});

export default connect(mapStateToProps, mapDispatchToProps)(EndGameModal);
