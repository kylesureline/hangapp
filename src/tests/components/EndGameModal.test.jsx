import React from 'react';
import { shallow } from 'enzyme';
import { EndGameModal } from '../../components/EndGameModal.jsx';
import { player } from '../fixtures/player';

let startSetPlayer, wrapper;

const playerWithFinishedGame = {
  ...player,
  guessesRemaining: 5,
  guessedWord: player.answer.word.split('')
};

beforeEach(() => {
  startSetPlayer = jest.fn();
  wrapper = shallow(
    <div id="app">
      <EndGameModal
        guessedWord={playerWithFinishedGame.guessedWord}
        guessesRemaining={playerWithFinishedGame.guessesRemaining}
        answer={playerWithFinishedGame.answer}
        startSetPlayer={startSetPlayer}
      />
    </div>
  );
});

test('should render EndGameModal correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

// need to deeply render?
// test('should startSetPlayer if closed', () => {
//   wrapper.find('#close-modal').simulate('click');
//   expect(startSetPlayer).toHaveBeenCalled();
// });
