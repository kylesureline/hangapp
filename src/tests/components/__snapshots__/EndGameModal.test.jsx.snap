// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EndGameModal correctly 1`] = `
<div
  id="app"
>
  <EndGameModal
    answer={
      Object {
        "def": "to cause to move with quick little jerks or oscillating motions",
        "type": "verb",
        "word": "jiggle",
      }
    }
    guessedWord={
      Array [
        "j",
        "i",
        "g",
        "g",
        "l",
        "e",
      ]
    }
    guessesRemaining={5}
    startSetPlayer={[MockFunction]}
  />
</div>
`;
